testcontainers/testcontainers-dotnet
# support identitytoken based authentication

This form of authentication is typically used by private registries which use OAuth2 authentication, for example Azure Container Registry (ACR). Specifically this PR fixes the situation where the identity token is given directly in the Docker `config.json` file.


## What does this PR do?

Proposes a fix for [issue 841](https://github.com/testcontainers/testcontainers-dotnet/issues/841)

Note: It is outside the scope of testcontainers library to _obtain_ the token.

Ref :  [Docker's documentation](https://docs.docker.com/engine/api/v1.42/#section/Authentication)

## Why is it important?

Support for Azure Container Registry from CI pipeline.

## Related issues

Relates #841

## How to test this PR

1. Have the Azure CLI installed.
1. Use a private Azure Container Registry. By private is meant one that doesn't allow un-authenticated pulls.
1. Login to Azure with an identity which is allowed to pull from that registry: `az login`
1. Perform implicit docker login:  `az acr login -name myprivateregistry`. This obtains a 3 hour token and uses the `docker` CLI to write the identitytoken into your local Docker config file. If you are doing this on a host which has a credstore available then "unfortunately" the Docker CLI is clever enough to use that. This is not good, as then you'll be testing something which testcontainers-dotnet already knows how to handle. The workaround for that is to do as follows:
```bash
TOKEN=$(az acr login --name myprivateregistry --expose-token --output tsv --query accessToken)
docker login myprivateregistry.azurecr.io --username 00000000-0000-0000-0000-000000000000 --password $TOKEN
```
4. Look in your `~/.docker/config.json` and verify that is has an explicit `identitytoken`.
5. Do the test.




## Implementation notes

This PR simply enhances on the  `Base64Provider` class. This is a little bit quick-and-dirty as technically identitytoken-based authentication has nothing to do with base64, so the name is slightly misleading for the identitytoken use-case. It may be "cleaner" to create a new provider class, e.g. `IdentityTokenProvider.cs`, although it would share 90% of its code with the existing `Base64Provider`.
